Mathlib/LinearAlgebra/LinearDisjoint.lean

```diff
@@ -0,0 +1,619 @@
+  `Submodule.LinearDisjoint.of_le_of_flat_left`, `Submodule.LinearDisjoint.of_le_of_flat_right`:
+  linearly disjoint is preserved by taking submodules under some flatness conditions.
+
+- `Submodule.LinearDisjoint.of_linearDisjoint_fg_left`,
+  `Submodule.LinearDisjoint.of_linearDisjoint_fg_right`,
+  `Submodule.LinearDisjoint.of_linearDisjoint_fg`:
+  conversely, if any finitely generated submodules of `M` and `N` are linearly disjoint,
+  then `M` and `N` themselves are linearly disjoint.
+
+- `Submodule.LinearDisjoint.bot_left`, `Submodule.LinearDisjoint.bot_right`:
+  the zero module is linearly disjoint with any other submodules.
+
+- `Submodule.LinearDisjoint.one_left`, `Submodule.LinearDisjoint.one_right`:
+  the image of `R` in `S` is linearly disjoint with any other submodules.
+
+- `Submodule.LinearDisjoint.of_left_le_one_of_flat`,
+  `Submodule.LinearDisjoint.of_right_le_one_of_flat`:
+  if a submodule is contained in the image of `R` in `S`, then it is linearly disjoint with
+  any other submodules, under some flatness conditions.
+
+- `Submodule.LinearDisjoint.not_linearIndependent_pair_of_commute_of_flat`,
+  `Submodule.LinearDisjoint.rank_inf_le_one_of_commute_of_flat`:
+  if `M` and `N` are linearly disjoint, if one of `M` and `N` is flat, then any two commutative
+  elements contained in the intersection of `M` and `N` are not `R`-linearly independent (namely,
+  their span is not `R ^ 2`). In particular, if any two elements in the intersection of `M` and `N`
+  are commutative, then the rank of the intersection of `M` and `N` is at most one.
+
+  These results are stated using bundled version (i.e. `a : ↥(M ⊓ N)`). If you want a not bundled
+  version (i.e. `a : S` with `ha : a ∈ M ⊓ N`), you may use `LinearIndependent.of_comp` and
+  `FinVec.map_eq` (in `Mathlib/Data/Fin/Tuple/Reflection.lean`),
+  see the following code snippet:
+
+  ```
+  have h := H.not_linearIndependent_pair_of_commute_of_flat hf ⟨a, ha⟩ ⟨b, hb⟩ hc
+  contrapose! h
+  refine .of_comp (M ⊓ N).subtype ?_
+  convert h
+  exact (FinVec.map_eq _ _).symm
+  ```
+
+- `Submodule.LinearDisjoint.rank_le_one_of_commute_of_flat_of_self`:
+  if `M` and itself are linearly disjoint, if `M` is flat, if any two elements in `M`
+  are commutative, then the rank of `M` is at most one.
+
+The results with name containing "of_commute" also have corresponding specified versions
+assuming `S` is commutative.
+
+## Tags
+
+linearly disjoint, linearly independent, tensor product
+
+-/
+
+open scoped Classical TensorProduct
+
+noncomputable section
+
+universe u v w
+
+namespace Submodule
+
+variable {R : Type u} {S : Type v}
+
+section Semiring
+
+variable [CommSemiring R] [Semiring S] [Algebra R S]
+
+variable (M N : Submodule R S)
+
+/-- Two submodules `M` and `N` in an algebra `S` over `R` are linearly disjoint if the natural map
+`M ⊗[R] N →ₗ[R] S` induced by multiplication in `S` is injective. -/
+@[mk_iff]
+protected structure LinearDisjoint : Prop where
```

Comment:
Why `protected`?

Reply:
Because later we will have `Subalgebra.LinearDisjoint`, `IntermediateField.LinearDisjoint` and `Field.LinearDisjoint`.